crates/shared/src/maintenance.rs

```diff
@@ -32,36 +43,56 @@ impl Maintaining for ServiceMaintenance {
 }
 
 impl ServiceMaintenance {
-    async fn run_maintenance_for_block_stream(self, block_stream: impl Stream<Item = Block>) {
-        futures::pin_mut!(block_stream);
-
+    pub async fn run_maintenance_for_blocks(self, blocks: impl Stream<Item = Block>) {
         let metrics = Metrics::instance(global_metrics::get_metric_storage_registry()).unwrap();
+        for label in ["success", "failure"] {
+            metrics.runs.with_label_values(&[label]).reset();
+        }
+
+        let blocks = blocks.fuse();
+        futures::pin_mut!(blocks);
 
-        while let Some(block) = block_stream.next().await {
+        let mut retry_block = None;
+
+        while let Some(block) = match retry_block.take() {
+            // We have a pending retry to process. First see if there is a new
+            // block that becomes available within a certain retry delay, and if
+            // there is, prefer that over the old outdated block.
+            Some(block) => time::timeout(self.retry_delay, blocks.next())
+                .await
+                .unwrap_or(Some(block)),
+            None => blocks.next().await,
+        } {
             tracing::debug!(
                 "running maintenance on block number {:?} hash {:?}",
                 block.number,
-                block.hash
+                block.hash,
             );
 
-            let block = block.number.unwrap_or_default().as_u64();
-            metrics.last_seen_block.set(block as _);
+            let block_number = block.number.unwrap_or_default().as_u64();
+
+            metrics.last_seen_block.set(block_number as _);
 
             if self
                 .run_maintenance()
-                .instrument(tracing::debug_span!("maintenance", block))
+                .instrument(tracing::debug_span!("maintenance", block = block_number))
                 .await
-                .is_ok()
+                .is_err()
             {
-                metrics.last_updated_block.set(block as _);
+                metrics.runs.with_label_values(&["failure"]).inc();
```

Comment:
Should this maybe also log that the maintenance failed? Otherwise logs might be confusing if you see multiple updates on the same block without any indication of a problem.